internal/access/destinationcredential.go

```diff
@@ -0,0 +1,195 @@
+	listener, err := data.ListenForNotify(c, rCtx.DataDB, data.ListenForNotifyOptions{
+		OrgID:                      rCtx.Authenticated.AccessKey.OrganizationID,
+		DestinationCredentialsByID: cr.ID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ListenForNotify: %w", err)
+	}
+	defer func() {
+		err := listener.Release(c)
+		if err != nil {
+			logging.Errorf("error releasing listener: %s", err)
+		}
+	}()
+
+	err = listener.WaitForNotification(c)
+	if err != nil {
+		logging.Debugf("CreateDestinationCredential notification timeout")
+		return nil, fmt.Errorf("WaitForNotification %w: %s", api.ErrTimeout, err)
+	}
+	logging.Debugf("CreateDestinationCredential notification received")
+
+	tx, err := rCtx.DataDB.Begin(rCtx.Request.Context(), &sql.TxOptions{
+		Isolation: sql.LevelReadCommitted,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("n: %w", err)
+	}
+	tx = tx.WithOrgID(cr.OrganizationID)
+
+	// looks like the DestinationCredential has been updated; reload the destination credential
+	logging.Debugf("Reloading DestinationCredential with id %d, orgID %d", cr.ID, cr.OrganizationID)
+	cr, err = data.GetDestinationCredential(tx, cr.ID, cr.OrganizationID)
+	if err != nil {
+		return nil, fmt.Errorf("GetDestinationCredential: %w", err)
+	}
+	if err = tx.Rollback(); err != nil {
+		logging.Debugf("could not roll back transaction")
+	}
+
+	return cr, nil
+}
+
+func ListDestinationCredentials(c *gin.Context, destination string, lastUpdateIndex int64) (ListDestinationCredentialResponse, error) {
+	rCtx := GetRequestContext(c)
+
+	roles := []string{models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole}
+	_, err := RequireInfraRole(c, roles...)
+	err = HandleAuthErr(err, "grants", "list", roles...)
+	if err != nil {
+		return ListDestinationCredentialResponse{}, err
+	}
+
+	dest, err := data.GetDestination(rCtx.DBTxn, data.GetDestinationOptions{ByName: destination})
+	if err != nil {
+		return ListDestinationCredentialResponse{}, err
+	}
+
+	if lastUpdateIndex == 0 {
+		logging.Debugf("listing all destination credentials")
+		result, err := data.ListDestinationCredentials(rCtx.DBTxn, dest.ID)
+		if err != nil {
+			return ListDestinationCredentialResponse{}, err
+		}
+		if len(result) > 0 {
+			return ListDestinationCredentialResponse{Items: result}, err
+		}
+	}
+
+	// Close the request scoped txn to avoid long-running transactions.
+	if err := rCtx.DBTxn.Rollback(); err != nil {
+		return ListDestinationCredentialResponse{}, err
+	}
+
+	listenOpts := data.ListenForNotifyOptions{
+		DestinationCredentialsByDestinationID: dest.ID,
+		OrgID:                                 rCtx.DBTxn.OrganizationID(),
+	}
+	query := func() (ListDestinationCredentialResponse, error) {
+		return listDestinationCredentialsWithMaxUpdateIndex(rCtx, dest.ID)
+	}
+	return blockingRequest(rCtx, listenOpts, query, lastUpdateIndex)
+}
+
+type ListDestinationCredentialRespct struct {
```

Comment:
```suggestion
type ListDestinationCredentialResp struct {
```

Not sure, but this looked like a typo